lib/discover_endpoint.js

```diff
@@ -0,0 +1,305 @@
+}
+
+/**
+ * If api call gets invalid endoint exception, SDK should attempt to remove the invalid
+ * endpoint from cache.
+ * @api private
+ */
+function invalidateCachedEndpoints(response) {
+  var error = response.error;
+  var httpResponse = response.httpResponse;
+  if (error &&
+    (error.code === 'InvalidEndpointException' || httpResponse.statusCode === 421)
+  ) {
+    var request = response.request;
+    var operations = request.service.api.operations || {};
+    var inputShape = operations[request.operation] ? operations[request.operation].input : undefined;
+    var identifiers = marshallCustomeIdentifiers(request, inputShape);
+    var cacheKey = getCacheKey(request);
+    cacheKey = util.update(cacheKey, identifiers);
+    AWS.endpointCache.remove(cacheKey);
+  }
+}
+
+/**
+ * If endpoint is explicitly configured, SDK should not do endpoint discovery in anytime.
+ * @param [object] clien Service client object.
+ * @api private
+ */
+function hasCustomEndpoint(client) {
+  //if set endpoint is set for specific client, enable endpoint discovery will raise an error.
+  if (client._originalConfig && client._originalConfig.endpoint && client._originalConfig.endpointDiscoveryEnabled === true) {
+    throw util.error(new Error(), {
+      code: 'ConfigurationException',
+      message: 'If custome endpoint is supplied, endpointDiscoverEnabled should be turned off.'
+    });
+  };
+  var svcConfig = AWS.config[client.serviceIdentifier] || {};
+  return AWS.config.endpoint || svcConfig.endpoint || (client._originalConfig && client._originalConfig.endpoint);
+}
+
+/**
+ * @api private
+ */
+function isFalsey(value) {
+  return ['false', '0'].indexOf(value) >= 0;
+}
+
+/**
+ * If endpoint discovery should perform for this request when endpoint discovery is optional.
+ * SDK performs config resolution in order like below:
+ * 1. If turned on client configuration(default to off) then turn on endpoint discovery.
+ * 2. If turned on in env AWS_ENABLE_ENDPOINT_DISCOVERY then turn on endpoint discovery.
+ * 3. If turned on in shared ini config file with key 'endpoint_discovery_enabled', then
+ *   turn on endpoint discovery.
+ * @param [object] request request object.
+ * @api private
+ */
+function isEndpointDiscoveryApplicable(request) {
+  var service = request.service || {};
+  if (service.config.endpointDiscoveryEnabled === true) return true;
+  if (Object.prototype.hasOwnProperty.call(process.env, endpointDiscoveryEnabledEnv)) {
+    if (process.env[endpointDiscoveryEnabledEnv] === undefined) {
+      throw util.error(new Error(), {
+        code: 'ConfigurationException',
+        message: 'environmental variable AWS_ENABLE_ENDPOINT_DISCOVERY cannot be set to nothing'
+      });
+    }
+    if (!isFalsey(process.env[endpointDiscoveryEnabledEnv])) return true;
+  }
+  //shared ini file is only available in Node
+  if (util.isBrowser()) return false;
+
+  var configFile = {};
+  try {
+    var iniLoader = require('./shared-ini').iniLoader;
```

Comment:
We don't want to import this file directly as it will add bloat to the browser SDK.